platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/io/ParallelOutputStream.kt

```diff
@@ -0,0 +1,248 @@
+        writeByteArray(b, off, len)
+    }
+
+    override fun write(b: Int) {
+        buffer.put(b.toByte())
+        maybeFlush()
+    }
+
+    override fun close() {
+        // send remaining data
+        if (buffer.position() > 0) {
+            sendBuffer()
+            takeNextBuffer()
+        }
+        // send a last empty buffer to signal the end
+        sendBuffer()
+        onClose()
+        buffers.rethrowFailureIfAny()
+        super.close()
+    }
+
+    private
+    tailrec fun writeByteArray(b: ByteArray, off: Int, len: Int) {
+        if (len <= 0) {
+            return
+        }
+        val remaining = buffer.remaining()
+        if (remaining > len) {
+            putByteArrayAndFlush(b, off, len)
+        } else {
+            putByteArrayAndFlush(b, off, remaining)
+            writeByteArray(b, off + remaining, len - remaining)
+        }
+    }
+
+    private
+    fun putByteArrayAndFlush(b: ByteArray, off: Int, len: Int) {
+        buffer.put(b, off, len)
+        maybeFlush()
+    }
+
+    private
+    fun maybeFlush() {
+        if (!buffer.hasRemaining()) {
+            sendBuffer()
+            takeNextBuffer()
+        }
+    }
+
+    private
+    fun sendBuffer() {
+        buffer.flip()
+        ready.offer(buffer)
+    }
+
+    private
+    fun takeNextBuffer() {
+        buffer = buffers.take()
+    }
+}
+
+
+/**
+ * Manages a pool of buffers of fixed [capacity][ByteBufferPool.bufferCapacity] allocated on-demand
+ * upto a [fixed maximum][ByteBufferPool.bufferCount].
+ */
+private
+class ByteBufferPool {
+
+    companion object {
+
+        /**
+         * How many bytes are transferred, at a time, from the producer to the writer thread.
+         *
+         * The smaller the number the more parallelism between producer and writer.
+         * The default is `32` for increased parallelism.
+         */
+        val bufferCapacity = System.getProperty("org.gradle.configuration-cache.internal.buffer-capacity", null)?.toInt()
+            ?: 32
+
+        /**
+         * Maximum number of buffers to be allocated.
+         *
+         * Determines the maximum memory working set: [bufferCount] * [bufferCapacity].
+         * The default maximum working set is `32MB`.
+         */
+        val bufferCount = System.getProperty("org.gradle.configuration-cache.internal.buffer-count", null)?.toInt()
+            ?: (1024 * 1024)
+
+        val timeoutMinutes: Long = System.getProperty("org.gradle.configuration-cache.internal.buffer-timeout-minutes", null)?.toLong()
+            ?: 30L /* stream can be kept open during the whole configuration phase */
+    }
+
+    private
+    val buffers = ArrayBlockingQueue<ByteBuffer>(bufferCount)
+
+    private
+    var buffersToAllocate = buffers.remainingCapacity()
+
+    private
+    val failure = AtomicReference<Exception>(null)
+
+    fun put(buffer: ByteBuffer) {
+        buffer.flip()
+        if (!buffers.offer(buffer, timeoutMinutes, TimeUnit.MINUTES)) {
```

Comment:
🤔  Can we actually block here? As far as I can see, the queue capacity is enough to fit all buffers. So the timeout can be an `assert`.

Reply:
Well spotted, Mike. We can simply assert on `offer`.